### PR TITLE
Add extract_ipv6_gua function

### PR DESCRIPTION
**Pull request by dkerr64**
Adds a new function `extract_ipv6_gua()` that is required for IPv6 support.

### CODE REVIEW

**Comment by dkerr64** [comment]
> Once PR #161 is merged would you rebase this and add some tests? See PR #175 for an example you can follow.

Yes I will work on that.

**Comment by dkerr64** [comment]
Ready to merge after #198 merged first.

**Comment by dkerr64** [comment]
@rhansen this one is ready for your review and merge (unless you find errors).

**Comment by dkerr64** [comment]
OMG... I could have been overthinking this.  As the only difference between any old IPv6 and a GUA IPv6 is that the address must start 2xxx: or 3xxx: then the regex can get simplified to....

```
my $regex_ipv6_gua = qr/
    (?=[23][A-F0-9]{3}:)$regex_ipv6
/xi;
```

Now isn't that nice. @rhansen

**Comment by rhansen** [comment]
That is nice.

Rather than a positive match on 2000::/3, I would rather reject addresses in ranges that are known to not be globally routable (multicast, link-local, ULA, documentation/example, etc.). Right now IANA is only assigning from 2000::/3, but that could theoretically change at any point.

**Comment by dkerr64** [comment]
> That is nice.
> 
> Rather than a positive match on 2000::/3, I would rather reject addresses in ranges that are known to not be globally routable (multicast, link-local, ULA, documentation/example, etc.). Right now IANA is only assigning from 2000::/3, but that could theoretically change at any point.

Well, you're right but that complicates things somewhat. I can exclude multicast, link local (LLA) and unique local (ULA) fairly easily...
```
my $regex_ipv6_gua = qr/
    (?!F[CDF][A-F0-9]{2}:)(?!FE[89A-F][A-F0-9])$regex_ipv6
/xi;
```
But we also need to exclude `::/128` and `::1/128` which means matching possibly any number of zeros separated by colons or one double colon.  There is probably a elegant regex for that which will take some thinking.

Oh... and all the test cases will need updated as right now they assume 2000::/3 are the only valid GUAs.

But [RFC 4291](https://tools.ietf.org/html/rfc4291) is clear... _"everything else"_ is a GUA unless otherwise documented (and ULAs would be a case of that) and [RFC 3587](https://tools.ietf.org/html/rfc3587) states _"Even though currently only 2000::/3 is being delegated by the IANA, implementations should not make any assumptions about 2000::/3 being special.  In the future, the IANA might be directed to delegate currently unassigned portions of the IPv6 address space for the purpose of Global Unicast as well."_ So we do have to figure it out.

**Comment by dkerr64** [comment]
> But we also need to exclude `::/128` and `::1/128` which means matching possibly any number of zeros separated by colons or one double colon. There is probably a elegant regex for that which will take some thinking.

Thinking about this further and reading the RFCs, a GUA address will always have a global routing prefix, namely a /64 or more likely a /48 prefix that will not be zero.  However looking at how things have been defined to-date... with the largest prefix assigned as a /10 for link-local I think it highly  likely that all future GUAs assigned will have a non-zero /16 prefix.  Put another way I propose...
```
my $regex_ipv6_gua = qr/
    (?!F[CDF][A-F0-9]{2}:)(?!FE[89A-F][A-F0-9])(?=[A-F0-9]{1,4}:)$regex_ipv6
/xi;
```
So if it is not in the loopback / link local / ULA range then it must still have at least some value in the first 16-bit segment of the address in order to be a GUA.  So anything starting `::` is eliminated.  The above regex allows for `0000:` which we should also eliminate.  But my point is that testing beyond the most significant 16 bits is I think unnecessary. @rhansen

**Comment by rhansen** [comment]
> a GUA address will always have a global routing prefix, namely a /64 or more likely a /48 prefix that will not be zero.

I don't think that's a safe assumption to make. In particular, I can imagine globally routable IPv4-mapped IPv6 addresses with the first 64 bits all zeros.

In general, it's better to err on the side of accepting a non-global address than to err on the side of rejecting a global address. And we only really need to filter out the non-global addresses that are likely to appear in the wild. Specifically, I'm OK with just filtering out:
  * unique local
  * link-local
  * exactly ::1 (don't worry about alternative representations)
  * exactly :: (don't worry about alternative representations)

For completeness, here are the non-global addresses I am aware of (see [RFC6890 section 2.2.3](https://tools.ietf.org/html/rfc6890#section-2.2.3)):
  * ::/128 [RFC4291](https://tools.ietf.org/html/rfc4291) unspecified address
  * ::1/128 [RFC4291](https://tools.ietf.org/html/rfc4291) loopback address
  * ::ffff:0:0/96 [RFC4291](https://tools.ietf.org/html/rfc4291) IPv4-mapped address
  * 100::/64 [RFC6666](https://tools.ietf.org/html/rfc6666) discard-only
  * 2001::/23 [RFC7343](https://tools.ietf.org/html/rfc7343) ORCHIDv2
  * 2001:2::/48 [RFC5180](https://tools.ietf.org/html/rfc5180) benchmarking
  * 2001:db8::/32 [RFC3849](https://tools.ietf.org/html/rfc3849) documentation
  * fc00::/7 [RFC4193](https://tools.ietf.org/html/rfc4193) unique local
  * fe80::/10 [RFC4291](https://tools.ietf.org/html/rfc4291) link-local
  * ff00::/8 [RFC4291](https://tools.ietf.org/html/rfc4291) multicast

File: t/is-and-extract-ipv6-gua.pl [hunk 1]
```diff
@@ -0,0 +1,580 @@
+use Test::More;
+use B qw(perlstring);
+
+SKIP: { eval { require Test::Warnings; } or skip($@, 1); }
+eval { require 'ddclient'; } or BAIL_OUT($@);
+
+
+my @valid_ipv6_not_gua = (
```
Comment (rhansen) [inline t/is-and-extract-ipv6-gua.pl:8]: There is too much redundancy with `t/is-and-extract-ipv6.pl`. I would drastically simplify by testing only:
  * the last address before each rejected block
  * the first address of each rejected block
  * the last address of each rejected block
  * the first address after each rejected block

Comment (dkerr64) [reply t/is-and-extract-ipv6-gua.pl:8]: I will update the tests

**Comment by dkerr64** [comment]
> > a GUA address will always have a global routing prefix, namely a /64 or more likely a /48 prefix that will not be zero.
> 
> I don't think that's a safe assumption to make. In particular, I can imagine globally routable IPv4-mapped IPv6 addresses with the first 64 bits all zeros.

IPv4-mapped IPv6 (::ffff:0:0/32) are not globally routable.  They exist only within a system network stack to allow an application to be written as IPv6 only but still talk to other network devices that are IPv4.  The system network stack translates from one to the other such that all traffic over the network is IPv4 while the application sees a IPv4-mapped IPv6.

All globally routable IPv6 addresses must have a global routing prefix in the most significant bits.  RFCs recommend that IANA assign between /48 and /64 global routing prefixes to sites.  More to the point, for our purposes we are checking only that the IPv6 address makes sense to register with DNS.  It will never be valid to register an address that has all zeros as the global routing prefix... as that would be non-routable.

So the question is how many zeros to check for.  I assert that it is good enough to check that /16 is not all zero which is already far more expansive than the previous 2000::/3 check.

As for the other defined non-global prefixes (documentation, benchmarking, etc.) I don't think we should bother checking and prohibiting them.  We don't do it for the IPv4 equivalents and it just feels a step too far.

**Comment by rhansen** [comment]
> > > a GUA address will always have a global routing prefix, namely a /64 or more likely a /48 prefix that will not be zero.
> > 
> > I don't think that's a safe assumption to make. In particular, I can imagine globally routable IPv4-mapped IPv6 addresses with the first 64 bits all zeros.
> 
> IPv4-mapped IPv6 (::ffff:0:0/32) are not globally routable.

By "IPv4-mapped IPv6 address" I was referring to the general idea of embedding an IPv4 address in an IPv6 address, not ::ffff:0:0/96 specifically. I was simply trying to state that some future RFC might reserve a globally routable block of addresses where the first 64 bits are all zeros, and that the most likely user of such a block would be an IPv4 transition protocol.

> All globally routable IPv6 addresses must have a global routing prefix in the most significant bits.

Perhaps that's true now, but it could change. I agree that it's unlikely to change, but why bake in that assumption when we don't have to? As I said before, we only need to filter out the non-global addresses we are likely to see in the wild:

  * unique local
  * link-local
  * multicast
  * the string `'::1'` (don't worry about alternative representations of ::1/128 such as `'::0:1'`)
  * the string `'::'` (don't worry about alternative representations of ::/128 such as `'::0'`)

File: ddclient.in [hunk 2]
```diff
@@ -2370,6 +2369,42 @@ sub extract_ipv6 {
...
+    (?!F[CDF][A-F0-9]{2}:)  # Not a multicast or ULA
+    (?!FE[89AB][A-F0-9]:)   # And not a link local
+    (?!0{0,4}:)             # And does not start with zeros
+    (?=[A-F0-9]{1,4}:)      # But does start with something
+    $regex_ipv6             # And is a valid IPv6 address
+/xi;
+
+######################################################################
+## is_ipv6_gua() validates if string is valid IPv6 Global Unicast Address (GUA)
+## with no preceding or trailing spaces/characters, not even line breaks.
```
Comment (rhansen) [inline ddclient.in:2392]: I don't like using the term GUA here because it has a specific meaning that no longer matches the behavior of these functions. How about:

```suggestion
## is_ipv6_global() returns true if the string is a valid IPv6 address that is
## likely to be globally routable, with no preceding or trailing characters.
```

File: ddclient.in [hunk 3]
```diff
@@ -2370,6 +2369,42 @@ sub extract_ipv6 {
...
+## start with a non-zero /16 prefix. The negative lookahead at the very
+## start is not required but significantly reduces the number of steps
+## required to match (ie, approx 2x faster).
+######################################################################
+my $regex_ipv6_gua = qr/
+    (?<![:.0-9A-Z])         # Not preceded by a character that could be in the address
+    (?!F[CDF][A-F0-9]{2}:)  # Not a multicast or ULA
+    (?!FE[89AB][A-F0-9]:)   # And not a link local
+    (?!0{0,4}:)             # And does not start with zeros
+    (?=[A-F0-9]{1,4}:)      # But does start with something
```
Comment (rhansen) [inline ddclient.in:2386]: Why is this line needed?

**Comment by dkerr64** [comment]
> Why is this line needed?

@rhansen Not sure why it won't let me quote reply, so copied comment manually... but... you're right.  There was logic that made me put it there but I just realized that the "does not start with zero" line before it had the same effect, so it can be removed.  But of course it has penalties... doubling again the time to perform the match.  Fiddling at [regex101](https://regex101.com/r/lNAbOJ/1) and commenting out this line and the first negative lookbehind makes the match at that page 5x slower (~2x contributed by commenting out the first lookahead, another ~2x from the "but does start with something").  If we don't care about performance (we're talking milliseconds here) then yes can take these lines out.

File: ddclient.in [hunk 4]
```diff
@@ -2370,6 +2369,39 @@ sub extract_ipv6 {
...
+## A Global Unicast Address (GUA) is any address in a range not otherwise
+## reserved.  So rule out any ULA (FC00::/7), multicast (FF00::/8) or
+## link local (FE80::/10) addresses. We also require that it start
+## with a non-zero /16 prefix though strictly a GUA could.
+######################################################################
+my $regex_ipv6_global = qr/
+    (?!FF[A-F0-9]{2}:)      # Not a multicast
+    (?!F[CD][A-F0-9]{2}:)   # And not a ULA
+    (?!FE[89AB][A-F0-9]:)   # And not a link local
+    (?!0{0,4}:)             # And does not start with zeros
```
[the remark this answers is not in the document]
Comment (rhansen) [reply ddclient.in]: > Perl is misbehaving with comments embedded within regex... cannot have anything that looks like a regex directive...

It's misbehaving because you're using `qr/.../`. Perl sees the slash in the comment as an end-of-regex marker. Try something like `qr{...}`, `qr@...@`, or `qr%...%` and it should work.

> simply testing for :: or ::1 is not good enough, you need 0000:/16

Why?

File: t/is-and-extract-ipv6-gua.pl [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+eval { require 'ddclient'; } or BAIL_OUT($@);
+
+my @valid_ipv6_not_global = (
+    # Unassigned address
+    "::",
+    "::0",
+    "::0:0",
+    "::0:0:0:0:0:0:0",
+    "0:0:0:0:0:0:0:0",
+    "0::0",
```
Comment (rhansen) [inline t/is-and-extract-ipv6-gua.pl:14]: Don't bother with these—we're not going to see them in the wild.

File: t/is-and-extract-ipv6-gua.pl [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+    # link local address
+    "::1",
+    "::0:0:0:01",
+    "::0:0:0:001",
+    "::0:0:0:0001",
+    "0::1",
+    "0:0::1",
+    "0:0:0:0:0:0:0:1",
```
Comment (rhansen) [inline t/is-and-extract-ipv6-gua.pl:22]: Ditto.

File: t/is-and-extract-ipv6-gua.pl [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+    "FC11::",
+    "FC11:2222::123.123.123.123",
+    "FD11:2222:3333:4444:5555:6666::8888",
+    "FD11:2222:3333:4444:5555::8888",
+    "FD11:2222::8888",
+    "FD11::3333:4444:5555:6666:123.123.123.123",
+    "FD11::4444:5555:6666:123.123.123.123",
+    "FD11::5555:6666:123.123.123.123",
+    "FD11::6666:123.123.123.123",
+    "FD11::8888",
```
Comment (rhansen) [inline t/is-and-extract-ipv6-gua.pl:43]: Just include the first and last address of the range:

```suggestion
    "fc00::",
    "fdff:ffff:ffff:ffff:ffff:ffff:ffff:ffff",
```

File: t/is-and-extract-ipv6-gua.pl [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+    # Link local unicast addresses
+    "FE81:2222:3333:4444:5555::7777:8888",
+    "FE81:2222:3333:4444::7777:8888",
+    "FE81::8888",
+    "FE91:2222:3333::7777:8888",
+    "FEA1:2222::7777:8888",
+    "FEB1::7777:8888",
```
Comment (rhansen) [inline t/is-and-extract-ipv6-gua.pl:50]: Ditto:

```suggestion
    "fe80::",
    "febf:ffff:ffff:ffff:ffff:ffff:ffff:ffff",
```

File: t/is-and-extract-ipv6-gua.pl [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+    # Multicast addresses
+    "FF11::",
+    "FF11:2222::",
+    "FF11:2222:3333::",
+    "FF11:2222:3333:4444::",
+    "FF11:2222:3333:4444:5555::",
+    "FF11:2222:3333:4444:5555:6666::",
+    "FF11:2222:3333:4444:5555:6666:7777::",
+    "FF11:2222:3333:4444:5555:6666:7777:8888",
```
Comment (rhansen) [inline t/is-and-extract-ipv6-gua.pl:59]: Ditto:

```suggestion
    "ff00::",
    "ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff",
```

File: t/is-and-extract-ipv6-gua.pl [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+    "2111:2222::6666:123.123.123.123",
+    "2111::3333:4444:5555:6666:123.123.123.123",
+    "2111::4444:5555:6666:123.123.123.123",
+    "2111::5555:6666:123.123.123.123",
+    "2111::6666:123.123.123.123",
+    "64:ff9b::123.123.123.123",
+    "64:ff9b:1::123.123.123.123",
+);
+
+my @invalid_ipv6 = (
```
Comment (rhansen) [inline t/is-and-extract-ipv6-gua.pl:108]: Delete this. Validity is already tested in `t/is-and-extract-ipv6.pl`.

File: t/is-and-extract-ipv6-gua.pl [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+    "FF11:2222::",
+    "FF11:2222:3333::",
+    "FF11:2222:3333:4444::",
+    "FF11:2222:3333:4444:5555::",
+    "FF11:2222:3333:4444:5555:6666::",
+    "FF11:2222:3333:4444:5555:6666:7777::",
+    "FF11:2222:3333:4444:5555:6666:7777:8888",
+);
+
+my @valid_ipv6_global = (
```
Comment (rhansen) [inline t/is-and-extract-ipv6-gua.pl:62]: Reduce this to addresses just outside the known-non-global blocks:

```perl
my @valid_ipv6_global = (
    "::2",
    "fbff:ffff:ffff:ffff:ffff:ffff:ffff:ffff",  # just before fc00::/7 ULA block
    "fe00::",                                   # just after  fc00::/7 ULA block
    "fe7f:ffff:ffff:ffff:ffff:ffff:ffff:ffff",  # just before fe80::/10 link-local block
    "fec0::",                                   # just after  fe80::/10 link-local block
    "feff:ffff:ffff:ffff:ffff:ffff:ffff:ffff",  # just before ff00::/8 multicast block
);
```

File: Makefile.am [hunk 12]
```diff
@@ -74,6 +74,7 @@ handwritten_tests = \
 	t/geturl_ssl.pl \
 	t/is-and-extract-ipv4.pl \
 	t/is-and-extract-ipv6.pl \
+	t/is-and-extract-ipv6-gua.pl \
```
Comment (rhansen) [inline Makefile.am:77]: Rename the file to `t/is-and-extract-ipv6-global.pl`.

File: ddclient.in (hunk 4 above)
[the remark this answers is not in the document]
Comment (dkerr64) [reply ddclient.in]: > > Perl is misbehaving with comments embedded within regex... cannot have anything that looks like a regex directive...
> 
> It's misbehaving because you're using `qr/.../`. Perl sees the slash in the comment as an end-of-regex marker. Try something like `qr{...}`, `qr@...@`, or `qr%...%` and it should work.

Thank you, that fixed it.

File: t/is-and-extract-ipv6-gua.pl (hunk 10 above)
Comment (dkerr64) [reply t/is-and-extract-ipv6-gua.pl:108]: That tests a different regex.  We need to make sure that the updated regex for global addresses doesn't break anything

File: Makefile.am (hunk 12 above)
Comment (dkerr64) [reply Makefile.am:77]: Done

File: t/is-and-extract-ipv6-gua.pl (hunk 7 above)
Comment (dkerr64) [reply t/is-and-extract-ipv6-gua.pl]: Done

File: t/is-and-extract-ipv6-gua.pl (hunk 9 above)
Comment (dkerr64) [reply t/is-and-extract-ipv6-gua.pl]: Done

File: t/is-and-extract-ipv6-gua.pl (hunk 11 above)
Comment (dkerr64) [reply t/is-and-extract-ipv6-gua.pl]: Done, but I have kept some of the other valid IPv6 globals because I don't think just testing on those "outside the range" is valid test.  I don't think we will ever see globally routable addresses like these. Anything in the f000::/4 range will probably the the last assigned.  I did not include ::2 as it is not globally routable.

**Comment by rhansen** [comment]
I pushed a bunch of commits to your branch. Please take a look and let me know what you think. Feel free to squash any or all of them into your commit. (Don't worry about crediting me.)

**Comment by dkerr64** [comment]
> I pushed a bunch of commits to your branch. Please take a look and let me know what you think. Feel free to squash any or all of them into your commit. (Don't worry about crediting me.)

I'm fine with all your changes and rebased/squashed them all.

I notice you moved some test data into `t/lib/ddclient/t.pm` which is good idea.  I'd like to introduce you to some work I have underway to create a `get_ip_from_interface` and `get_default_interface` function (properly implements the request in PR #50).  I have been collecting a lot more output from ip / ifconfig / netstat and currently have it [here](https://github.com/dkerr64/ddclient/blob/IPv6/t/get_ip_from_if.pl).  It looks like all these should move into the new file you created.  The functions btw are currently [here](https://github.com/dkerr64/ddclient/blob/3c87b6b16d8709fca3557a5f0d19ed2b7d1af762/ddclient.in#L2859-L2992).

**Comment by dkerr64** [comment]
Thank you @rhansen
